staging/vhost-device-gpu/src/device.rs

```diff
@@ -0,0 +1,1639 @@
+                GuestMemoryMmap::<()>::from_ranges(&[(GuestAddress(0), 0x1000)]).unwrap(),
+            );
+            backend.update_memory(mem.clone()).unwrap();
+
+            let vring = VringRwLock::new(mem, 0x1000).unwrap();
+            vring.set_queue_info(0x100, 0x200, 0x300).unwrap();
+            vring.set_queue_ready(true);
+
+            assert_eq!(
+                backend
+                    .handle_event(0, EventSet::OUT, &[vring.clone()], 0)
+                    .unwrap_err()
+                    .kind(),
+                io::ErrorKind::Other
+            );
+
+            assert_eq!(
+                backend
+                    .handle_event(1, EventSet::IN, &[vring.clone()], 0)
+                    .unwrap_err()
+                    .kind(),
+                io::ErrorKind::Other
+            );
+
+            // Hit the loop part
+            backend.set_event_idx(true);
+            backend
+                .handle_event(0, EventSet::IN, &[vring.clone()], 0)
+                .unwrap();
+
+            // Hit the non-loop part
+            backend.set_event_idx(false);
+            backend.handle_event(0, EventSet::IN, &[vring], 0).unwrap();
+        }
+    }
+
+    mod test_image {
+        use super::*;
+        const GREEN_PIXEL: u32 = 0x00FF00FF;
+        const RED_PIXEL: u32 = 0xFF0000FF;
+        const BYTES_PER_PIXEL: usize = 4;
+
+        pub fn write(mem: &GuestMemoryMmap, image_addr: GuestAddress, width: u32, height: u32) {
+            let mut image_addr: u64 = image_addr.0;
+            for i in 0..width * height {
+                let pixel = if i % 2 == 0 { RED_PIXEL } else { GREEN_PIXEL };
+                let pixel = pixel.to_be_bytes();
+
+                mem.memory()
+                    .write_slice(&pixel, GuestAddress(image_addr))
+                    .unwrap();
+                image_addr += BYTES_PER_PIXEL as u64;
+            }
+        }
+
+        pub fn assert(data: &[u8], width: u32, height: u32) {
+            assert_eq!(data.len(), (width * height) as usize * BYTES_PER_PIXEL);
+            for (i, pixel) in data.chunks(BYTES_PER_PIXEL).enumerate() {
+                let expected_pixel = if i % 2 == 0 { RED_PIXEL } else { GREEN_PIXEL };
+                assert_eq!(
+                    pixel,
+                    expected_pixel.to_be_bytes(),
+                    "Wrong pixel at index {i}"
+                );
+            }
+        }
+    }
+
+    fn split_into_mem_entries(
+        addr: GuestAddress,
+        len: u32,
+        chunk_size: u32,
+    ) -> Vec<virtio_gpu_mem_entry> {
+        let mut entries = Vec::new();
+        let mut addr = addr.0;
+        let mut remaining = len;
+
+        while remaining >= chunk_size {
+            entries.push(virtio_gpu_mem_entry {
+                addr,
+                length: chunk_size,
+                padding: Default::default(),
+            });
+            addr += u64::from(chunk_size);
+            remaining -= chunk_size;
```

Comment:
Wouldn't this potentially overflow? Try this:

```suggestion
            remaining = remaining.saturating_sub(chunk_size);
```

This will saturate `remaining` to zero instead of overflowing.

Reply:
But how can it overflow (underflow)? The loop condition is `while remaining >= chunk_size {` ?

Reply:
Yes you're right, I misread it!